src/betree.h

```diff
@@ -0,0 +1,139 @@
+#ifndef BETREE_H__
+#define BETREE_H__
+
+#include <stdbool.h>
+#include <stdlib.h>
+
+struct pred {
+    unsigned int variable_id;
+    int value;
+};
+
+struct sub {
+    unsigned int id;
+    unsigned int variable_id_count;
+    // TODO Make const
+    unsigned int* variable_ids;
+    const struct ast_node *expr;
+};
+
+struct event {
+    unsigned int pred_count;
+    // TODO Make const
+    struct pred** preds;
+};
+
+struct cnode;
+
+struct lnode {
+    struct cnode* parent;
+    unsigned int sub_count;
+    // TODO Make const
+    struct sub **subs;
+    unsigned int max;
+};
+
+struct pdir;
```

Comment:
While I applaud the use of opaque struct, you seem to have declared some structs multiple time. Based on what I'm reading, it seems intentional as you declare the struct just about the struct that's about to use it. I'm on the fence about this but it's largely subjective and I can't come up with any counter-arguments so I'm happy either way.

Reply:
Oh wait they're not opaque structs, they're just declared below :P